pyspread/doc/pyspread.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH PYSPREAD 1 "February 06, 2013"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
pyspread \- A non-traditional Python spreadsheet application.
.SH SYNOPSIS
.B pyspread\fR [\fIoptions\fR] "\fIfiles\fR" ...
.SH DESCRIPTION
.B pyspread
is a non-traditional spreadsheet application that is based on and written in the
programming language Python.
The goal of pyspread is to be the most pythonic spreadsheet.
Pyspread expects Python expressions in its grid cells, which makes a spreadsheet
specific language obsolete. Each cell returns a Python object that can be accessed
from other cells. These objects can represent anything including lists or matrices.
.PP
.SH OPTIONS
This program follows the usual GNU command line syntax, with long
options starting with two dashes ('-').
.TP
.BR \-h\fR, " " \fB\-\-help
Show summary of options.
.TP
.BR \-\-version
Show version of program.
.TP
.BR \-d " " \fIDIMENSIONS\fB ", " \-\-dimensions=\fIDIMENSIONS\fR

Size of empty table, which is initially created.
.I
DIMENSIONS
has the format (Rows, Columns, Tables).
.I
DIMENSIONS
defaults to (1000, 100, 3).
.SH AUTHOR
pyspread was written by Martin Manns.
.PP
This manual page was written by Martin Manns <mmanns@gmx.net>,
for the Debian project (but may be used by others).


changelog
Changelog for pyspread
======================

Note: This changelog starts at version 0.2.0

0.2.4
-----

Fix some typos and offline tutorial
Pie charts added to chart_dialog
Charts fit better in figure
Multiline EntryLine implemented
Large bitmaps that are pasted from the clipboard are not scaled down any more
Legend support added to chart_dialog
troff file for man page updated
man page updated
Changed initial module order to gain Windows compatibility
TableChoiceIntCtrl is now in EntryLineToolbarPanel
Selection mode introduced, that is toggled by new ToggleButton in EntryLine
Copy results of figure now copies bitmap into clipboard
man page file added to setup.py
SVG export for matplotlib figures added
Old config files are now updated correctly (window geometry always and GPG key uid for pre 0.2.4 files)
Title added to charts in chart_dialog
The font choice widget now uses the system GUI default font as basis
Statusbar tooltip removed
Signing pys files made more robust
man page moved into the docs folder
Pyspread can now be started from other Python applications
Stock icons for Windows replaced by Tango icons
Find and Find/Replace icon added to Windows stock item replacement
Attributes toolbar now refreshes after update so that status changes are visible
GPG key choice dialog buttons have new labels
Cells with matplotlib figure can now also be exported to png, pdf, eps and ps
Safe mode status icon added to status bar
Copy results of numpy arrays now creates lists
Paste As menu item added
Paste As Dialog unrolls iterables to grid
Transpose operation added to Paste As dialog
Color attribute added to pie chart dialog
Chart grid added
Boxplots added to chart dialog
All columns in chart dialog are now growable
Histograms added to chart dialog
Some bug fixes on chart dialog
Toolbars switched to wx.lib.agw.aui to fix display bug
Merged cells are now redrawn correctly even when they are only partly visible
Python tutorial path upadated and moved to sysvars
Errors are now redirected to the shell on Windows (no extra window)
Pyspread now also starts if python-gupg is installed but gnupg is not
Testing via py.test can now be called via setup.py test
Behavior on invalid background in Windows fixed
Translations updated
Help now shows up on non UTF-8 systems (e. g. Windows)
Bug in csv import fixed by reverting to type comparisons
Batch file for Windows added
Bash shell script now keeps pre-defiend PYTHONPATH
Fixed bug that made ChartDialog ignore its code parameter
CartDialog now parses checkboxes correctly
Windows icon added, Windows bat and icon added to setup.py
Unit tests extended
runtests is now executable
Clipboard data from multiple cells can now be transposed
Paste as from multiple cells is now done with the correct type
Format string variables are changed to named variables to ease translation.
Error message removed when closing bitmap macro filechoice without choosing a file.
Chart dialog now supports dates as x axis labels
Dependency restriction to wxPython 2.8 added because 2.9.x has GUI issues
File approval warning made less severe
Tooltip labels added for plot attributes
Figure and axes parameter tooltips added to chart dialog
Chart dialog now supports annotations
Icon added to .desktop file
First step document adjusted to Jekyll
FAQ updated
Help html window now goes back with right click
UTF-8 symbols removed from tutorial
Assignment identification in model.py changed to ast
Year numbers removed from Copyright notices

0.2.3
-----

Pivot table example added
i18n.py added in lib directory
Metadata for pypi updated
Package info changed to Python 2.7 required
INSTALL file made version independent
setup.py modified for pip
All unit tests are now directory independent
Developer documentation extended
Merge cell functionailty added
Fixed bug that results in incomplete saves
Bitmap rendering added
Pasting bitmaps now works
TextCtrl length limit now stored in config
Bitmaps can now be copied as results
Bugfix: Passwd was stored even though user did opt for not storing it
Globals are now cleared when a new spreadsheet is created or opened
Initial globals state is now reset upon File->New
Modules are now reloaded when globals are cleared.
Matplotlib charts basically work
PEP 8 code polish for event.py
Chart dialog added
X-axis added to chart dialog
Pasting into selection implemented
Chart code generation added
gitignore file extended to ignore dist builds and the MANIFEST
Recursion limit increased to 10**6
Chart data now evaluates like grid cell
Matplotlib added to dependencies
Pasting is now undone in one step per pasting operation not in one step per cell
Fixed csv import dialog
Shellscript added to start pyspread from top directory
Situation handled when getdefaultlocale returns None in i18n.py
Signature files added to gitignore
Signature files in examples removed
Marker colors now work in chart dialog
Marker styles can now be set from chart dialog
Line styles now work in chart dialog
Charts are now correctly parsed when globals are referenced
Added how to start hacking to README
Two icons added to menubar; menubar code formatting
Macro toolbar with chart insert tool adeded
Pasting into EntryLine is now possible
Copying from EntryLine is now possible
Menus for image and image link insertion added
Axes panel added to chart dialog
Bar chart entry added to chart_dialog
xscale and yscale work in chart dialog
Fixed a bug when loading files with inconsistent row widths
Images and charts example added
Dependencies removed from example files
Danish translation file added
GPG key creation switched to python-gnupg
Passwords removed. GPG keys with passwords cannot be used any longer.
Removed pyme from gpg.py comments and from setup.py dependencies
Unit tests adapted for python-gnupg
Locale directory added
Dutch po file added and converted to mo file
Added a message when choosing GPG keys that a password must not be set
Help documents are updated for changes and new features

pyspread/src/pyspread.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Copyright Martin Manns
# Distributed under the terms of the GNU General Public License

# --------------------------------------------------------------------
# pyspread is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# pyspread is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with pyspread.  If not, see <http://www.gnu.org/licenses/>.
# --------------------------------------------------------------------

"""

========
pyspread
========

Python spreadsheet application

Run this script to start the application.

Provides
--------

* Commandlineparser: Gets command line options and parameters
* MainApplication: Initial command line operations and application launch

"""

import sys
import optparse

import wx
__ = wx.App(False)  # Windows Hack

from sysvars import get_program_path
import lib.i18n as i18n


#use ugettext instead of getttext to avoid unicode errors
_ = i18n.language.ugettext

sys.setrecursionlimit(10000)
sys.path.insert(0, get_program_path())

# Patch for using with PyScripter thanks to Colin J. Williams
# If wx exists in sys,modules, we dont need to import wx version.
# wx is already imported if the PyScripter wx engine is used.

try:
    sys.modules['wx']
except KeyError:
    # Select wx version 2.8 if possible
    try:
        import wxversion
        wxversion.select(['2.8', '2.9'])

    except ImportError:
        pass

from src.gui._events import post_command_event, GridActionEventMixin

DEBUG = False


class Commandlineparser(object):
    """
    Command line handling

    Methods:
    --------

    parse: Returns command line options and arguments as 2-tuple

    """

    def __init__(self):
        from src.config import config
        self.config = config

        usage_str = _("usage: %prog [options] [filename]")
        version = config["version"]
        version_str = _("%prog {version}").format(version=version)
        self.parser = optparse.OptionParser(usage=usage_str,
                                            version=version_str)

        grid_shape = (
            config["grid_rows"],
            config["grid_columns"],
            config["grid_tables"],
        )

        self.parser.add_option(
            "-d", "--dimensions", type="int", nargs=3,
            dest="dimensions", default=grid_shape,
            help=_("Dimensions of empty grid (works only without filename) "
                   "rows, cols, tables [default: %default]")
        )

    def parse(self):
        """
        Returns a a tuple (options, filename)

        options: The command line options
        filename: String (defaults to None)
        \tThe name of the file that is loaded on start-up

        """
        options, args = self.parser.parse_args()

        # If one dimension is 0 then the grid has no cells
        if min(options.dimensions) < 1:
            print _("Cell dimension must be > 0.")
            sys.exit()

        # No MDI yet, pyspread can be started several times though
        if len(args) > 1:
            print _("Only one file may be opened at a time.")
            sys.exit()

        filename = None
        if len(args) == 1:
            # A filename is provided and hence opened
            filename = args[0]

        return options, filename

# end of class Commandlineparser


class MainApplication(wx.App, GridActionEventMixin):
    """Main application class for pyspread."""

    dimensions = (1, 1, 1)  # Will be overridden anyways
    options = {}
    filename = None

    def __init__(self, *args, **kwargs):

        try:
            self.S = kwargs.pop("S")
        except KeyError:
            self.S = None

        # call parent class initializer
        wx.App.__init__(self, *args, **kwargs)

    def OnInit(self):
        """Init class that is automatically run on __init__"""

        # Get command line options and arguments
        self.get_cmd_args()

        # Initialize the prerequisitions to construct the main window
        wx.InitAllImageHandlers()

        # Main window creation
        from src.gui._main_window import MainWindow

        self.main_window = MainWindow(None, title="pyspread", S=self.S)

        ## Initialize file loading via event

        # Create GPG key if not present

        try:
            from src.lib.gpg import genkey
            genkey()

        except ImportError:
            pass

        except ValueError:
            # python-gnupg is installed but gnupg is not insatlled

            pass

        # Show application window
        self.SetTopWindow(self.main_window)
        self.main_window.Show()

        # Load filename if provided
        if self.filepath is not None:
            post_command_event(self.main_window, self.GridActionOpenMsg,
                               attr={"filepath": self.filepath})
            self.main_window.filepath = self.filepath

        return True

    def get_cmd_args(self):
        """Returns command line arguments

        Created attributes
        ------------------

        options: dict
        \tCommand line options
        dimensions: Three tuple of Int
        \tGrid dimensions, default value (1,1,1).
        filename: String
        \tFile name that is loaded on start

        """

        cmdp = Commandlineparser()
        self.options, self.filepath = cmdp.parse()

        if self.filename is None:
            rows, columns, tables = self.options.dimensions
            cmdp.config["grid_rows"] = str(rows)
            cmdp.config["grid_columns"] = str(columns)
            cmdp.config["grid_tables"] = str(tables)


def pyspread(S=None):
    """Parses command line and starts pyspread"""

    # Initialize main application
    app = MainApplication(S=S, redirect=False)

    app.MainLoop()


if __name__ == "__main__":
    if 'unicode' not in wx.PlatformInfo:
        print _("You need a unicode build of wxPython to run pyspread.")

    else:
        if DEBUG:
            import cProfile
            cProfile.run('pyspread()')
        else:
            pyspread()